src/cli/self_update/test.rs

```diff
@@ -1,57 +1,85 @@
 //! Support for functional tests.
 
-use std::sync::Mutex;
+use std::{io, sync::Mutex};
 
-#[cfg(not(unix))]
+#[cfg(windows)]
 use winreg::{
     enums::{HKEY_CURRENT_USER, KEY_READ, KEY_WRITE},
     RegKey, RegValue,
 };
 
-#[cfg(not(unix))]
-pub fn get_path() -> std::io::Result<Option<RegValue>> {
-    let root = RegKey::predef(HKEY_CURRENT_USER);
-    let environment = root
-        .open_subkey_with_flags("Environment", KEY_READ | KEY_WRITE)
-        .unwrap();
-    match environment.get_raw_value("PATH") {
-        Ok(val) => Ok(Some(val)),
-        Err(ref e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
-        Err(e) => Err(e),
-    }
-}
-
-#[cfg(not(unix))]
-fn restore_path(p: Option<RegValue>) {
-    let root = RegKey::predef(HKEY_CURRENT_USER);
-    let environment = root
-        .open_subkey_with_flags("Environment", KEY_READ | KEY_WRITE)
-        .unwrap();
-    if let Some(p) = p.as_ref() {
-        environment.set_raw_value("PATH", p).unwrap();
-    } else {
-        let _ = environment.delete_value("PATH");
-    }
-}
-
-/// Support testing of code that mutates global path state
-pub fn with_saved_path(f: &mut dyn FnMut()) {
+/// Support testing of code that mutates global state
+fn with_saved_global_state<S>(
+    getter: impl Fn() -> io::Result<Option<S>>,
+    setter: impl Fn(Option<S>),
+    f: &mut dyn FnMut(),
+) {
     // Lock protects concurrent mutation of registry
     static LOCK: Mutex<()> = Mutex::new(());
     let _g = LOCK.lock();
 
-    // On windows these tests mess with the user's PATH. Save
-    // and restore them here to keep from trashing things.
-    let saved_path = get_path().expect("Error getting PATH: Better abort to avoid trashing it.");
-    let _g = scopeguard::guard(saved_path, restore_path);
+    // Save and restore the global state here to keep from trashing things.
+    let saved_state =
+        getter().expect("Error getting global state: Better abort to avoid trashing it");
+    let _g = scopeguard::guard(saved_state, setter);
 
     f();
 }
 
+pub fn with_saved_path(f: &mut dyn FnMut()) {
+    with_saved_global_state(get_path, restore_path, f)
+}
+
+#[cfg(windows)]
+pub fn get_path() -> io::Result<Option<RegValue>> {
+    get_reg_value(&RegKey::predef(HKEY_CURRENT_USER), "Environment", "PATH")
+}
+
 #[cfg(unix)]
-pub fn get_path() -> std::io::Result<Option<()>> {
+pub fn get_path() -> io::Result<Option<()>> {
     Ok(None)
 }
 
+#[cfg(windows)]
+fn restore_path(p: Option<RegValue>) {
+    restore_reg_value(&RegKey::predef(HKEY_CURRENT_USER), "Environment", "PATH", p)
+}
+
 #[cfg(unix)]
 fn restore_path(_: Option<()>) {}
+
+#[cfg(windows)]
+pub fn with_saved_programs_display_version(f: &mut dyn FnMut()) {
```

Comment:
Looking back, it could be that this is overly specific for `rustup`'s public API. We could possibly expose `with_saved_reg_value(root, subkey, name)` instead, and `with_saved_programs_display_version()` can then be easily defined on the tests side.

Reply:
This is in the test API anyway, so I don't see any issue.